gk/fib.c

```diff
@@ -0,0 +1,1337 @@
+		 * Ethernet header entry from the neighbor hash table.
+		 */
+		if (gw_addr->proto == ETHER_TYPE_IPv4 && gw_addr->ip.v4.s_addr
+				== ip_prefix_addr.ip.v4.s_addr) {
+			/*
+			 * The Ethernet header cache entry is still
+			 * referenced by another prefix, so we cannot
+			 * release this FIB entry.
+			 */
+			if (eth_cache->ref_cnt != 0) {
+				RTE_LOG(ERR, GATEKEEPER,
+					"gk: delete a gateway fib entry [ip prefix = %s], however, its cached Ethernet header is referenced by other entries\n",
+					ip_prefix);
+				return -1;
+			}
+
+			if (ip_prefix_fib->action == GK_FWD_GATEWAY_FRONT_NET)
+				neigh_fib = gk_conf->neigh_fib_front;
+			else
+				neigh_fib = gk_conf->neigh_fib_back;
+
+			neigh_ht = &neigh_fib->u.neigh;
+			ret = rte_hash_del_key(neigh_ht->hash_table,
+				&gw_addr->ip.v4.s_addr);
+			if (ret < 0) {
+				RTE_LOG(ERR, GATEKEEPER,
+					"gk: failed to delete the Ethernet cached header for the IP prefix %s\n",
+					ip_prefix);
+				return -1;
+			}
+
+			/*
+			 * TODO Add concurrency control to
+			 * reset the Ethernet cache entry.
+			 */
+			eth_cache->stale = false;
+			memset(&eth_cache->eth_hdr, 0,
+				sizeof(eth_cache->eth_hdr));
+			eth_cache->ref_cnt = 0;
+		} else if (gw_addr->proto == ETHER_TYPE_IPv6 &&
+				memcmp(gw_addr->ip.v6.s6_addr,
+				ip_prefix_addr.ip.v6.s6_addr,
+				sizeof(gw_addr->ip.v6) == 0)) {
+			/*
+			 * The Ethernet header cache entry is still
+			 * referenced by another prefix, so we cannot
+			 * release this FIB entry.
+			 */
+			if (eth_cache->ref_cnt != 0) {
+				RTE_LOG(ERR, GATEKEEPER,
+					"gk: delete a gateway fib entry [ip prefix = %s], however, its cached Ethernet header is referenced by other entries\n",
+					ip_prefix);
+				return -1;
+			}
+
+			if (ip_prefix_fib->action == GK_FWD_GATEWAY_FRONT_NET)
+				neigh_fib = gk_conf->neigh6_fib_front;
+			else
+				neigh_fib = gk_conf->neigh6_fib_back;
+
+			neigh_ht = &neigh_fib->u.neigh6;
+			ret = rte_hash_del_key(neigh_ht->hash_table,
+				gw_addr->ip.v6.s6_addr);
+			if (ret < 0) {
+				RTE_LOG(ERR, GATEKEEPER,
+					"gk: failed to delete the Ethernet cached header for the IP prefix %s\n",
+					ip_prefix);
+				return -1;
+			}
+
+			/*
+			 * TODO Add concurrency control to
+			 * reset the Ethernet cache entry.
+			 */
+			eth_cache->stale = false;
+			memset(&eth_cache->eth_hdr, 0,
+				sizeof(eth_cache->eth_hdr));
+			eth_cache->ref_cnt = 0;
+		}
+
+		break;
+	}
+
+	case GK_DROP:
+		break;
+
+	case GK_FWD_NEIGHBOR_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_NEIGHBOR_BACK_NET:
+		/* FALLTHROUGH */
+	default:
+		return -1;
+		break;
+	}
+
+	/*
+	 * If the updated FIB entry doesn't have an action GK_FWD_GRANTOR,
+	 * we don't need to flush the flow tables at all.
+	 */
+	if (ip_prefix_fib->action == GK_FWD_GRANTOR) {
+
+		rte_atomic16_init(&ip_prefix_fib->num_updated_instances);
+
+		/* Send the FIB entry to the GK mailboxes. */
+		for (i = 0; i < gk_conf->num_lcores; i++) {
+			notify_flow_flush(ip_prefix_fib,
```

Comment:
Yes, but I recommend two changes:

1. The current loop should keep track of which instances received the notification, so if at least one instance did not, the loop can repeat over the missing instances; the loop would repeat for at most three times. It'd be simpler to just keep trying to notify the same instance three times, that is, adding an inner loop to the current loop, but this would be bad because it would give too little time for the failing instance to process its mailbox.

2. If after three iterations of the main loop there's still instances that did not receive the notification, a log entry must be issued, and the waiting loop next should wait for only the instances that received the notification.